File: 2022/02/python/main.py
#!/usr/bin/env python

from dataclasses import dataclass
from enum import Enum
from os.path import dirname, join

# Day 02

# Common


def read_input(filename):
    data = join(dirname(__file__), '..', filename)
    with open(data) as f:
        return f.read().strip()


test = read_input('test.txt')
data = read_input('input.txt')


MAPPING = {
    'A': 1,
    'B': 2,
    'C': 3,
    'X': 1,  # Rock of loose
    'Y': 2,  # Paper or draw
    'Z': 3,  # Scissor or win
}

# Bitwise precalculated
CASES = {
    1: (4, 3),
    2: (3, 5),
    3: (5, 4),
}


@dataclass
class Match:
    opponent: int
    you: int

    def score(self):
        # Draw is 3 points
        if self.opponent == self.you:
            return self.you + 3

        won = CASES[self.you][0] == self.you + self.opponent
        return self.you + (6 if won else 0)


@dataclass
class Game:
    matches: list[Match]

    def play(self):
        return sum([match.score() for match in self.matches])

    # Part 2
    def cheat(self):
        for match in self.matches:
            # Draw
            if match.you == 2:
                match.you = match.opponent
            else:
                win = 0 if match.you == 1 else 1
                match.you = CASES[match.opponent][win] - match.opponent

    @staticmethod
    def parse(data: str):
        lines = [line.split(' ') for line in data.split('\n')]
        matches = [Match(MAPPING[line[0]], MAPPING[line[1]]) for line in lines]
        return Game(matches)


# 1
print('1.')
game = Game.parse(test)
print(game.play())
game = Game.parse(data)
print(game.play())


# 2
print('\n2.')
game = Game.parse(test)
game.cheat()
print(game.play())
game = Game.parse(data)
game.cheat()
print(game.play())

File: 2022/02/rust/main.rs
#![feature(test)]

extern crate test;

#[cfg(test)]
mod tests {
    use super::*;
    use test::Bencher;

    #[bench]
    fn bench_a(b: &mut Bencher) {
        b.iter(|| part_a(INPUT));
    }

    #[bench]
    fn bench_a_old(b: &mut Bencher) {
        b.iter(|| part_a_old(INPUT));
    }

    #[bench]
    fn bench_b(b: &mut Bencher) {
        b.iter(|| part_b(INPUT));
    }
}

const TEST: &str = include_str!("../test.txt");
const INPUT: &str = include_str!("../input.txt");

#[derive(Debug, Clone, Copy)]
enum Type {
    Rock = 1,
    Paper = 2,
    Scissor = 3,
}

#[derive(Debug)]
struct Game(Type, Type);

impl Game {
    fn score(&self) -> u32 {
        let mut total = self.0 as u32;
        // Result will be: 0=draw 1=win 2=loss
        let result = (self.0 as i8 - self.1 as i8).rem_euclid(3) as u8;
        total += match result {
            1 => 6,
            0 => 3,
            _ => 0,
        };
        total
    }
}

fn part_a_old(input: &str) {
    let result = input
        .trim()
        .lines()
        .map(|line| {
            let opponent: Type = match line.chars().nth(0).unwrap() {
                'A' => Type::Rock,
                'B' => Type::Paper,
                'C' => Type::Scissor,
                _ => panic!("unkown input"),
            };
            let me = match line.chars().nth(2).unwrap() {
                'X' => Type::Rock,
                'Y' => Type::Paper,
                'Z' => Type::Scissor,
                _ => panic!("unkown input"),
            };
            let g: Game = Game(me, opponent);
            g.score()
        })
        .sum::<u32>();
    println!("A: {:?}", result);
}

fn part_a(input: &str) {
    let result = input
        .trim()
        .lines()
        .map(|line| {
            let bytes = line.as_bytes();
            let opponent = bytes[0] - b'A';
            let me: u8 = bytes[2] - b'X';

            let total = me + 1;
            // Result will be: 0=draw 1=win 2=loss
            let result = (me as i8 - opponent as i8).rem_euclid(3);
            let score = match result {
                1 => 6,
                0 => 3,
                _ => 0,
            };
            (total + score) as u16
        })
        .sum::<u16>();
    println!("{}", result);
}

fn part_b(input: &str) {
    let result = input
        .trim()
        .lines()
        .map(|line| {
            let bytes = line.as_bytes();
            let opponent = bytes[0] - 'A' as u8;
            let result: u8 = bytes[2] - 'X' as u8;

            let total: u8 = match result {
                1 => 3,
                2 => 6,
                _ => 0,
            };
            let me = (opponent + result + 2).rem_euclid(3);
            (total + me + 1) as u16
        })
        .sum::<u16>();
    println!("{}", result);
}

fn main() {
    println!("Part A:");
    part_a(TEST);
    part_a(INPUT);

    println!("\nPart B:");
    part_b(TEST);
    part_b(INPUT);
}
